### ISSUE TITLE
B5: Segment-Aware Patch Enrichment

### ISSUE THREAD

**Issue by iremurhan** [iremurhan/tez_v2_clean#59]
## Motivation
 
CLIP's ViT encodes an image as a sequence of patch embeddings, then pools them into a single CLS vector. This pooling discards **which patches belong to the same object**. Two images with the same objects in different spatial arrangements can produce nearly identical CLS embeddings.
 
The idea: give the ViT an extra signal — a **segment embedding** — so patches that belong to the same object share an identity, analogous to how BERT uses segment embeddings to distinguish sentence A from sentence B.
 
## Pipeline Overview (Step by Step)
 
The full pipeline has two phases: an **offline preprocessing** phase (runs once) and an **online training** phase (runs every batch).
 
### Phase 1: Offline — Precompute SAM Masks
 
This runs **once** before training and produces a segment map file per image.
 
```
┌─────────────┐     ┌──────────────────┐     ┌─────────────────────────┐
│ Input image  │────▶│  Frozen SAM       │────▶│  Binary masks per       │
│ (COCO/Flickr)│     │  (auto mode)      │     │  segment (N segments)   │
└─────────────┘     └──────────────────┘     └───────────┬─────────────┘
                                                         │
                                                         ▼
                                              ┌─────────────────────────┐
                                              │  Segment map:           │
                                              │  H×W integer array      │
                                              │  pixel (i,j) → seg_id   │
                                              │  (0 = background)       │
                                              └───────────┬─────────────┘
                                                         │
                                                         ▼
                                              ┌─────────────────────────┐
                                              │  Save as .npy or .pt    │
                                              │  per image              │
                                              │  e.g., seg_maps/        │
                                              │    000000001234.npy      │
                                              └─────────────────────────┘
```
 
**Details:**
- Use SAM (Segment Anything Model) in **automatic mask generation** mode — no prompts needed.
- SAM returns a list of binary masks per image. Assign each mask a unique integer ID (1, 2, 3, ...). Overlapping regions: take the mask with highest predicted IoU score.
- Output: one integer array of shape `(H, W)` per image where each pixel has a segment ID.
- Store these arrays on disk indexed by image ID. Total storage: ~5–10 GB for COCO (118K images × ~1MB each, compressed).
 
**SAM model choice:** Use `sam_vit_b` (base) for speed. We don't need high-quality boundaries — we just need a rough "which patches go together" signal.
 
### Phase 2: Offline — Patch-to-Segment Mapping
 
ViT-L/14@336 divides a 336×336 image into **24×24 = 576 patches**, each covering a 14×14 pixel region.
 
```
┌─────────────────────────────┐     ┌─────────────────────────────┐
│  Segment map (H×W)          │     │  Patch grid (24×24 = 576)   │
│  (resized to 336×336)       │────▶│  For each patch:            │
│                             │     │    majority-vote seg_id     │
│                             │     │    from its 14×14 region    │
└─────────────────────────────┘     └──────────────┬──────────────┘
                                                   │
                                                   ▼
                                    ┌─────────────────────────────┐
                                    │  patch_seg_ids: tensor of   │
                                    │  shape (576,) where each    │
                                    │  entry is a segment ID      │
                                    │  e.g., [0, 0, 1, 1, 1, 2,  │
                                    │         2, 3, 3, 3, ...]    │
                                    └─────────────────────────────┘
```
 
**This can be precomputed and stored alongside the segment map**, or computed on-the-fly in the dataloader (it's just a resize + majority vote per patch, very fast).
 
### Phase 3: Online — Segment Embedding in the ViT
 
This is the actual model modification.
 
```
Standard CLIP ViT input:
    input_i = patch_embed_i + pos_embed_i        (for patch i)
 
Our modification:
    input_i = patch_embed_i + pos_embed_i + seg_embed[seg_id_i]
                                              ▲
                                              │
                                    Learned embedding table
                                    nn.Embedding(N_seg, d_model)
                                    N_seg = max segment IDs (e.g., 64 or 128)
                                    d_model = 1024 (ViT-L)
```
 
**Key design decisions:**
 
| Decision | Choice | Rationale |
|----------|--------|-----------|
| N_seg (max segments) | 64 or 128 | SAM rarely produces >50 segments per image. Use modular arithmetic if exceeded: `seg_id % N_seg` |
| Embedding init | Zeros | Start with no effect, let training learn the signal. Alternative: small random init |
| Where to add | Before first transformer layer | Same position as positional embeddings — this is the standard injection point |
| What to train | Only seg_embed + existing trainable params | Seg embedding table is tiny (~64 × 1024 = 65K params). Freeze everything else per B0+ config |
 
**Forward pass change (pseudocode):**
```python
# In the vision encoder forward():
 
# Existing:
x = self.patch_embed(images)            # (B, 576, 1024)
x = x + self.pos_embed                  # (B, 576, 1024)
 
# NEW: add segment embedding
seg_ids = batch["seg_ids"]              # (B, 576), integers in [0, N_seg)
seg_emb = self.seg_embedding(seg_ids)   # (B, 576, 1024)
x = x + seg_emb                         # (B, 576, 1024)
 
# Then continue as normal:
x = self.transformer(x)
cls = x[:, 0]  # CLS token
```
 
### Summary: What Changes Where
 
| Component | Change | Files affected |
|-----------|--------|----------------|
| Preprocessing | New script: run SAM, save segment maps | `scripts/precompute_sam.py` (new) |
| Dataloader | Load segment map, compute patch→seg_id, return in batch | `dataset.py` or equivalent |
| Model | Add `nn.Embedding(N_seg, d_model)`, add to patch input | `model.py` or equivalent |
| Config | New fields: `model.seg_embed_size`, `data.seg_map_dir` | `config.yaml` |
| Training loop | No change (seg embedding is part of the forward pass) | — |
 
## Registry Entry
 
```yaml
B5_seg:
  description: "B0+ + segment-aware patch enrichment (SAM masks, Variant B)"
  overrides:
    model.seg_embed_size: 64
    data.seg_map_dir: "precomputed/sam_masks/"
```
 
## Evaluation
 
- Compare against B0+ and B4 (classification loss) on COCO 5K and Flickr30K.
- Hypothesis: segment embedding should help more on COCO (complex multi-object scenes) than Flickr (simpler scenes).
- Also run SugarCrepe to check if spatial/relational reasoning improves.
 
## Dependencies & Risks
 
- SAM model weights need to be downloaded (~375 MB for `vit_b`).
- Precomputation time: ~4–6 hours for COCO on a single GPU (can be parallelized).
- Risk: if SAM segments don't align well with semantic objects at the patch grid resolution, the signal may be too noisy. Mitigation: visualize a few examples before full precomputation.
 
## Blocking
 
- Not blocked by Issue 1 (B4). Can be developed in parallel.
- But **should run after B4**, so we can compare all three: B0+ (baseline) vs. B4 (classification signal) vs. B5 (segment signal).
